delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/auth/AuthTokenProvider.scala

```diff
@@ -0,0 +1,74 @@
+package ch.epfl.bluebrain.nexus.delta.sdk.auth
+
+import cats.effect.Clock
+import ch.epfl.bluebrain.nexus.delta.kernel.cache.KeyValueStore
+import ch.epfl.bluebrain.nexus.delta.kernel.utils.IOUtils
+import ch.epfl.bluebrain.nexus.delta.sdk.auth.AuthMethod.{Anonymous, Credentials}
+import ch.epfl.bluebrain.nexus.delta.sdk.identities.model.AuthToken
+import monix.bio.UIO
+
+import java.time.{Duration, Instant}
+
+/**
+  * Provides an auth token for the service account, for use when comunicating with remote storage
+  */
+trait AuthTokenProvider {
+  def apply(): UIO[Option[AuthToken]]
+}
+
+object AuthTokenProvider {
+  def apply(auth: Option[AuthMethod], keycloakAuthService: KeycloakAuthService): AuthTokenProvider = {
+    auth match {
+      case Some(credentials: Credentials)    => new CachingKeycloakAuthTokenProvider(credentials, keycloakAuthService)
+      case Some(AuthMethod.AuthToken(token)) => new FixedAuthTokenProvider(AuthToken(token))
+      case Some(Anonymous) | None            => new AnonymousAuthTokenProvider
+    }
+  }
+  def anonymousForTest: AuthTokenProvider = new AnonymousAuthTokenProvider
+}
+
+private class AnonymousAuthTokenProvider extends AuthTokenProvider {
+  override def apply(): UIO[Option[AuthToken]] = UIO.pure(None)
+}
+
+/**
+  * Uses a fixed (probably long-living) auth token. Should be removed when we are confident with the credentials method
+  */
+private class FixedAuthTokenProvider(authToken: AuthToken) extends AuthTokenProvider {
+  override def apply(): UIO[Option[AuthToken]] = UIO.pure(Some(authToken))
+}
+
+/**
+  * Uses the supplied credentials to get an auth token from keycloak. This token is cached until near-expiry to speed up
+  * operations
+  */
+private class CachingKeycloakAuthTokenProvider(credentials: Credentials, service: KeycloakAuthService)(implicit
+    clock: Clock[UIO]
+) extends AuthTokenProvider {
+  private val cache = KeyValueStore.create[Unit, AccessTokenWithMetadata]()
+
+  override def apply(): UIO[Option[AuthToken]] = {
+    for {
+      existingValue <- cache.get(())
+      now           <- IOUtils.instant
+      finalValue    <- existingValue match {
+                         case None                                 => fetchValue
+                         case Some(value) if isExpired(value, now) => fetchValue
```

Comment:
It would be nice to have some log whenever we fetch a new token

Reply:
We can debug log, do you want the value of the token logged?